lib/srv/discovery/access_graph_azure.go

```diff
@@ -0,0 +1,390 @@
+	}
+	features := azure_sync.BuildFeatures(supportedKinds...)
+
+	// Cancels the context to stop the event watcher if the access graph connection fails
+	var wg sync.WaitGroup
+	defer wg.Wait()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer cancel()
+		if !accessGraphConn.WaitForStateChange(ctx, connectivity.Ready) {
+			s.Log.InfoContext(ctx, "access graph service connection was closed")
+		}
+	}()
+
+	// Reconciles the resources as they're imported from Azure
+	azureResources := &azure_sync.Resources{}
+	ticker := time.NewTicker(15 * time.Minute)
+	defer ticker.Stop()
+	for {
+		err := s.reconcileAccessGraphAzure(ctx, azureResources, stream, features)
+		if errors.Is(err, errNoAccessGraphFetchers) {
+			err := stream.CloseSend()
+			if errors.Is(err, io.EOF) {
+				err = nil
+			}
+			return trace.Wrap(err)
+		}
+		select {
+		case <-ctx.Done():
+			return trace.Wrap(ctx.Err())
+		case <-ticker.C:
+		case <-reloadCh:
+		}
+	}
+}
+
+// initTAGAzureWatchers initializes the TAG Azure watchers
+func (s *Server) initTAGAzureWatchers(ctx context.Context, cfg *Config) error {
+	staticFetchers, err := s.accessGraphAzureFetchersFromMatchers(cfg.Matchers, "" /* discoveryConfigName */)
+	if err != nil {
+		s.Log.ErrorContext(ctx, "Error initializing access graph fetchers", "error", err)
+	}
+	s.staticTAGAzureFetchers = staticFetchers
+	if !cfg.AccessGraphConfig.Enabled {
+		return nil
+	}
+	go func() {
+		reloadCh := s.newDiscoveryConfigChangedSub()
+		for {
+			fetchers := s.getAllTAGSyncAzureFetchers()
+			// Wait for the config to change and re-evaluate the fetchers before starting the sync.
+			if len(fetchers) == 0 {
+				s.Log.DebugContext(ctx, "No Azure sync fetchers configured. Access graph sync will not be enabled.")
+				select {
+				case <-ctx.Done():
+					return
+				case <-reloadCh:
+					// if the config changes, we need to get the updated list of fetchers
+				}
+				continue
+			}
+			// Reset the Azure resources to force a full sync
+			if err := s.initializeAndWatchAzureAccessGraph(ctx, reloadCh); errors.Is(err, errTAGFeatureNotEnabled) {
+				s.Log.WarnContext(ctx, "Access Graph specified in config, but the license does not include Teleport Policy. Access graph sync will not be enabled.")
+				break
+			} else if err != nil {
+				s.Log.WarnContext(ctx, "Error initializing and watching access graph", "error", err)
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(time.Minute):
+			}
+		}
+	}()
+	return nil
+}
+
+func (s *Server) accessGraphAzureFetchersFromMatchers(matchers Matchers, discoveryConfigName string) ([]*azure_sync.Fetcher, error) {
+	var fetchers []*azure_sync.Fetcher
+	var errs []error
+	if matchers.AccessGraph == nil {
+		return fetchers, nil
+	}
+	for _, matcher := range matchers.AccessGraph.Azure {
+		fetcherCfg := azure_sync.Config{
+			CloudClients:        s.CloudClients,
+			SubscriptionID:      matcher.SubscriptionID,
```

Comment:
should we support wildcards?

Reply:
Yes

Reply:
https://github.com/gravitational/access-graph/issues/1376